[cache] CacheBasicClientTest.cacheManagerOfCache_cannotBeOverwrittenConcurrently

```
java.lang.AssertionError: Unexpected exception thrown: A cache named the-cache already exists.
	at com.hazelcast.cache.CacheBasicAbstractTest.cacheManagerOfCache_cannotBeOverwrittenConcurrently(CacheBasicAbstractTest.java:949)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:785)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast-client/725/testReport/junit/com.hazelcast.client.cache/CacheBasicClientTest/cacheManagerOfCache_cannotBeOverwrittenConcurrently/

--- a/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheHelper.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheHelper.java
@@ -37,7 +37,6 @@ import com.hazelcast.util.FutureUtil;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 
@@ -98,27 +97,27 @@ final class ClientCacheHelper {
     }
 
     /**
-     * Sends the cache configuration to server.
+     * Creates a new cache configuration on Hazelcast members.
      *
      * @param client             the client instance which will send the operation to server
-     * @param configs            {@link ConcurrentMap} based store that holds cache configurations
-     * @param cacheName          full cache name with prefixes
+     * @param existingConfig     an existing {@link CacheConfig}, already known to the client
      * @param newCacheConfig     the cache configuration to be sent to server
-     * @param createAlsoOnOthers flag which represents whether cache config
-     *                           will be sent to other nodes by the target node
-     * @param syncCreate         flag which represents response will be waited from the server
+     * @param createAlsoOnOthers when {@code true} the {@code newCacheConfig}
+     *                           will be sent to all cluster members by the target member that receives the
+     *                           invocation
+     * @param syncCreate         when {@code true}, this call will block until response is received from the member. This does not
+     *                           imply that when this method exits the {@code CacheConfig} is already created on all members.
      * @param <K>                type of the key of the cache
      * @param <V>                type of the value of the cache
      * @return the created cache configuration
+     * @see com.hazelcast.cache.impl.operation.CacheCreateConfigOperation
      */
     static <K, V> CacheConfig<K, V> createCacheConfig(HazelcastClientInstanceImpl client,
-                                                      ConcurrentMap<String, CacheConfig> configs,
-                                                      String cacheName,
+                                                      CacheConfig<K, V> existingConfig,
                                                       CacheConfig<K, V> newCacheConfig,
                                                       boolean createAlsoOnOthers,
                                                       boolean syncCreate) {
         try {
-            CacheConfig<K, V> currentCacheConfig = configs.get(cacheName);
             int partitionId = client.getClientPartitionService().getPartitionId(newCacheConfig.getNameWithPrefix());
 
             Object resolvedConfig = resolveCacheConfig(client, newCacheConfig, partitionId);
@@ -132,7 +131,7 @@ final class ClientCacheHelper {
                 final Data data = CacheCreateConfigCodec.decodeResponse(response).response;
                 return resolveCacheConfig(client, clientInvocation, data);
             } else {
-                return currentCacheConfig;
+                return existingConfig;
             }
         } catch (Exception e) {
             throw rethrow(e);

--- a/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheProxyFactory.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheProxyFactory.java
@@ -26,6 +26,8 @@ import com.hazelcast.config.CacheConfig;
 import com.hazelcast.config.NearCacheConfig;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
+import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
@@ -58,6 +60,14 @@ public class ClientCacheProxyFactory extends ClientProxyFactoryWithContext {
 
     void removeCacheConfig(String cacheName) {
         configs.remove(cacheName);
+    }
+
+    CacheConfig getCacheConfig(String cacheName) {
+        return configs.get(cacheName);
+    }
+
+    Set<Map.Entry<String, CacheConfig>> configs() {
+        return configs.entrySet();
     }
 
     @SuppressFBWarnings("RV_RETURN_VALUE_OF_PUTIFABSENT_IGNORED")

--- a/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/HazelcastClientCacheManager.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/HazelcastClientCacheManager.java
@@ -32,8 +32,6 @@ import java.net.URI;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Properties;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
 
 import static com.hazelcast.util.ExceptionUtil.rethrow;
 import static com.hazelcast.util.Preconditions.checkNotNull;
@@ -47,7 +45,6 @@ public final class HazelcastClientCacheManager extends AbstractHazelcastCacheMan
 
     private final HazelcastClientInstanceImpl client;
     private final ClientCacheProxyFactory clientCacheProxyFactory;
-    private final ConcurrentMap<String, CacheConfig> configs = new ConcurrentHashMap<String, CacheConfig>();
 
     public HazelcastClientCacheManager(HazelcastClientCachingProvider cachingProvider, HazelcastInstance hazelcastInstance,
                                        URI uri, ClassLoader classLoader, Properties properties) {
@@ -92,13 +89,11 @@ public final class HazelcastClientCacheManager extends AbstractHazelcastCacheMan
     @SuppressFBWarnings("RV_RETURN_VALUE_OF_PUTIFABSENT_IGNORED")
     @Override
     protected <K, V> void addCacheConfigIfAbsent(CacheConfig<K, V> cacheConfig) {
-        configs.putIfAbsent(cacheConfig.getNameWithPrefix(), cacheConfig);
         clientCacheProxyFactory.addCacheConfig(cacheConfig.getNameWithPrefix(), cacheConfig);
     }
 
     @Override
     protected void removeCacheConfigFromLocal(String cacheNameWithPrefix) {
-        configs.remove(cacheNameWithPrefix);
         clientCacheProxyFactory.removeCacheConfig(cacheNameWithPrefix);
     }
 
@@ -125,7 +120,7 @@ public final class HazelcastClientCacheManager extends AbstractHazelcastCacheMan
     @Override
     protected <K, V> CacheConfig<K, V> findCacheConfig(String cacheName, String simpleCacheName, boolean createAlsoOnOthers,
                                                        boolean syncCreate) {
-        CacheConfig<K, V> config = configs.get(cacheName);
+        CacheConfig<K, V> config = clientCacheProxyFactory.getCacheConfig(cacheName);
         if (config == null) {
             // if cache config not found, try to find it from partition
             config = getCacheConfig(cacheName, simpleCacheName);
@@ -140,7 +135,7 @@ public final class HazelcastClientCacheManager extends AbstractHazelcastCacheMan
     @Override
     protected <K, V> CacheConfig<K, V> createCacheConfig(String cacheName, CacheConfig<K, V> config,
                                                          boolean createAlsoOnOthers, boolean syncCreate) {
-        return ClientCacheHelper.createCacheConfig(client, configs, cacheName, config,
+        return ClientCacheHelper.createCacheConfig(client, clientCacheProxyFactory.getCacheConfig(cacheName), config,
                 createAlsoOnOthers, syncCreate);
     }
 
@@ -161,7 +156,7 @@ public final class HazelcastClientCacheManager extends AbstractHazelcastCacheMan
 
     @Override
     protected void postDestroy() {
-        Iterator<Map.Entry<String, CacheConfig>> iter = configs.entrySet().iterator();
+        Iterator<Map.Entry<String, CacheConfig>> iter = clientCacheProxyFactory.configs().iterator();
         while (iter.hasNext()) {
             Map.Entry<String, CacheConfig> entry = iter.next();
             String cacheName = entry.getKey();

--- a/hazelcast-client/src/test/java/com/hazelcast/client/cache/impl/ClientCacheHelperTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/cache/impl/ClientCacheHelperTest.java
@@ -119,21 +119,21 @@ public class ClientCacheHelperTest extends HazelcastTestSupport {
 
     @Test
     public void testCreateCacheConfig_whenSyncCreate_thenReturnNewConfig() {
-        CacheConfig<String, String> actualConfig = createCacheConfig(client, configs, CACHE_NAME, newCacheConfig, false, true);
+        CacheConfig<String, String> actualConfig = createCacheConfig(client, cacheConfig, newCacheConfig, false, true);
 
         assertNotEquals(cacheConfig, actualConfig);
     }
 
     @Test
     public void testCreateCacheConfig_whenNotSyncCreate_thenReturnCurrentConfig() {
-        CacheConfig<String, String> actualConfig = createCacheConfig(client, configs, CACHE_NAME, newCacheConfig, false, false);
+        CacheConfig<String, String> actualConfig = createCacheConfig(client, cacheConfig, newCacheConfig, false, false);
 
         assertEquals(cacheConfig, actualConfig);
     }
 
     @Test(expected = IllegalArgumentException.class)
     public void testCreateCacheConfig_rethrowsExceptions() {
-        createCacheConfig(exceptionThrowingClient, configs, CACHE_NAME, newCacheConfig, false, false);
+        createCacheConfig(exceptionThrowingClient, cacheConfig, newCacheConfig, false, false);
     }
 
     @Test

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCacheManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCacheManager.java
@@ -118,25 +118,24 @@ public abstract class AbstractHazelcastCacheManager
             throw new CacheException("A cache named " + cacheName + " already exists.");
         }
         // Create cache config on all nodes as sync
-        CacheConfig<K, V> currentCacheConfig = createCacheConfig(cacheName, newCacheConfig, true, true);
+        createCacheConfig(cacheName, newCacheConfig, true, true);
         // Create cache proxy object with cache config
         ICacheInternal<K, V> cacheProxy = createCacheProxy(newCacheConfig);
-        if (currentCacheConfig == null) {
-            // Put created cache config.
-            // Single thread region because "createConfigOnPartition" is single threaded by partition thread
-            addCacheConfigIfAbsent(newCacheConfig);
-            // Put created cache. No need to a "putIfAbsent" as this is a single threaded region
-            caches.put(newCacheConfig.getNameWithPrefix(), cacheProxy);
-            // Register listeners
+        // Add created cache config to local configurations map
+        addCacheConfigIfAbsent(newCacheConfig);
+        ICacheInternal<?, ?> existingCache = caches.putIfAbsent(newCacheConfig.getNameWithPrefix(), cacheProxy);
+        if (existingCache == null) {
+            // Register listeners on new cache
             registerListeners(newCacheConfig, cacheProxy);
             return cacheProxy;
+        } else {
+            CacheConfig config = existingCache.getConfiguration(CacheConfig.class);
+            if (config.equals(newCacheConfig)) {
+                return (ICacheInternal<K, V>) existingCache;
+            } else {
+                throw new CacheException("A cache named " + cacheName + " already exists.");
+            }
         }
-        ICacheInternal<?, ?> cache = getOrPutIfAbsent(currentCacheConfig.getNameWithPrefix(), cacheProxy);
-        CacheConfig config = cache.getConfiguration(CacheConfig.class);
-        if (config.equals(newCacheConfig)) {
-            return (ICacheInternal<K, V>) cache;
-        }
-        throw new CacheException("A cache named " + cacheName + " already exists.");
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheCreateConfigOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheCreateConfigOperation.java
@@ -46,6 +46,11 @@ import java.util.concurrent.atomic.AtomicInteger;
  * <p>
  * This operation's purpose is to pass the required parameters into
  * {@link ICacheService#putCacheConfigIfAbsent(com.hazelcast.config.CacheConfig)}.
+ * <p/>
+ * Caveat: This operation does not return a response when {@code createAlsoOnOthers} is {@code true} and there are at least one
+ * remote members on which the new {@code CacheConfig} needs to be created. When this operation returns {@code null} with {@code
+ * createAlsoOnOthers == true}, it is impossible to tell whether it is because the {@code CacheConfig} was not already registered
+ * or due to sending out the operation to other remote members.
  */
 public class CacheCreateConfigOperation
         extends AbstractNamedOperation

--- a/hazelcast/src/test/java/com/hazelcast/cache/CacheBasicSingleMemberTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cache/CacheBasicSingleMemberTest.java
@@ -1,0 +1,49 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class CacheBasicSingleMemberTest
+        extends CacheBasicAbstractTest {
+
+    TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
+    HazelcastInstance instance;
+
+    @Override
+    protected void onSetup() {
+        instance = factory.newHazelcastInstance(createConfig());
+    }
+
+    @Override
+    protected void onTearDown() {
+        factory.terminateAll();
+    }
+
+    @Override
+    protected HazelcastInstance getHazelcastInstance() {
+        return instance;
+    }
+}

